source/opt/spread_volatile_semantics.cpp

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/opt/spread_volatile_semantics.h"
+
+#include "source/opt/decoration_manager.h"
+#include "source/opt/ir_builder.h"
+
+namespace spvtools {
+namespace opt {
+namespace {
+
+const uint32_t kOpDecorateInOperandBuiltinDecoration = 2u;
+const uint32_t kOpLoadInOperandMemoryOperands = 1u;
+
+bool IsBuiltInForVolatileSemantics(uint32_t built_in) {
```

Comment:
Should this include the following?
```
case SpvBuiltinHelperInvocation:
  return module()->version() >= SPV_VERSION_WORD(1, 6);
```

That said, you'd probably want to ensure that OpDemoteToHelperInvocation is called in the entry point call tree.